### PR TITLE
Azure: add CPU and Memory cost approximations

### PR DESCRIPTION
**Pull request by logyball**
# Updates

Adding the CPU and Memory metric for azure.

> [!IMPORTANT]  
> The CPU and Memory ratio is cribbed directly from [AWS](https://github.com/grafana/cloudcost-exporter/blob/46620104ee4a4c1858e4f1ec33f6ac570cf90485/pkg/aws/ec2/pricing_map.go#L39).  Under testing, the numbers add up, so 🤷 

Calculating price with the ratios of CPU to Memory.  Adding the CPU and Memory metrics to AKS exporter.

Updated the PriceMap from the retail price SDK SKU object to a custom object `MachineSku`, that simply contains the price information (that's the only thing we care about).

### Other things

- adding `family` label

# Testing

## General Purpose

example searching for a particular node name:

```sh
❯ curl -s localhost:8080/metrics | grep "<name redacted>"

cloudcost_azure_aks_instance_cpu_usd_per_core_hour{cluster_name="<cluster-name>",instance="<instance-name>",machine_type="Standard_D16s_v3",operating_system="Linux",price_tier="OnDemand",region="<region>"} 0.039
cloudcost_azure_aks_instance_memory_usd_per_gib_hour{cluster_name="<cluster-name>"instance="<instance-name>",machine_type="Standard_D16s_v3",operating_system="Linux",price_tier="OnDemand",region="<region>"} 0.0052499999999999995
cloudcost_azure_aks_instance_total_usd_per_hour{cluster_name="<cluster-name>",instance="<instance-name>",machine_type="Standard_D16s_v3",operating_system="Linux",price_tier="OnDemand",region="<region>"} 0.96
```

Azure Pricing:
image

`Standard_D16s_v3` stats:
- vCPU: 16
- RAM: 64 GiB

(64 *  0.0052499999999999995) + (0.039 * 16) = 0.336 + 0.624 = 0.96

And for another:

```sh
❯  curl -s localhost:8080/metrics | grep "<name redacted>" 
cloudcost_azure_aks_instance_cpu_usd_per_core_hour{machine_type="Standard_D8s_v3",operating_system="Linux",price_tier="OnDemand"} 0.039
cloudcost_azure_aks_instance_memory_usd_per_gib_hour{machine_type="Standard_D8s_v3",operating_system="Linux",price_tier="OnDemand"} 0.0052499999999999995
cloudcost_azure_aks_instance_total_usd_per_hour{machine_type="Standard_D8s_v3",operating_system="Linux",price_tier="OnDemand"} 0.48
```

image

`Standard_D8s_v3` stats:
- vCPU: 8
- RAM: 32 GiB

(32 *  0.0052499999999999995) + (0.039 * 8) = 0.168 + 0.312 = 0.48

## Memory Optimized

```sh
❯ curl -s localhost:8080/metrics | grep "<name redacted>"
cloudcost_azure_aks_instance_cpu_usd_per_core_hour{instance_family="Memory optimized",machine_type="Standard_E4ds_v4",operating_system="Linux",price_tier="OnDemand"} 0.041519999999999994
cloudcost_azure_aks_instance_memory_usd_per_gib_hour{instance_family="Memory optimized",machine_type="Standard_E4ds_v4",operating_system="Linux",price_tier="OnDemand"} 0.0056225
cloudcost_azure_aks_instance_total_usd_per_hour{instance_family="Memory optimized",machine_type="Standard_E4ds_v4",operating_system="Linux",price_tier="OnDemand"} 0.346
```

image

`Standard_E4ds_v4` stats:
- vCPU: 4
- RAM: 32 GiB

(32 *  0.0056225) + (0.041519999999999994 * 4) = 0.17992 + 0.16608 = 0.346

### CODE REVIEW

File: pkg/azure/aks/machine_store.go [hunk 1]
```diff
@@ -56,9 +56,9 @@ type VirtualMachineInfo struct {
 	OperatingSystem MachineOperatingSystem
 	Priority        MachinePriority
 
-	NumOfCores     int32
-	MemoryInMiB    int32
-	OsDiskSizeInMB int32
+	NumOfCores     float64
+	MemoryInMiB    float64 // Note, the Azure Docs say MiB, the golang docs say MB, we're going with the Azure Docs :nervous:
```
Comment (the-it) [inline pkg/azure/aks/machine_store.go:60]: It's pretty easy to spot this, when we introduce it to the cost metrics ;-) ... we will find out

File: pkg/azure/aks/aks.go [hunk 2]
```diff
@@ -63,19 +63,19 @@ var (
 	InstanceCPUHourlyCostDesc = prometheus.NewDesc(
 		prometheus.BuildFQName(cloudcost_exporter.MetricPrefix, subsystem, "instance_cpu_usd_per_core_hour"),
 		"The cpu cost a compute instance in USD/(core*h)",
-		[]string{"instance", "region", "machine_type", "cluster_name", "price_tier", "operating_system"},
```
Comment (the-it) [inline pkg/azure/aks/aks.go:66]: I had a chat (live in person, I know the world is crazy) about the `price_tier` topic with @bastischubert. Some of our customers are very eager to get metrics about this (which OpenCost isn't doing). Maybe you can reach out to and we can find out if this attribute is already enough, or if they need something else.

**Review by the-it — APPROVED** [review]
lgtm ... left comments (more for organizational purposes)

File: pkg/azure/aks/price_store.go [hunk 3]
```diff
@@ -16,13 +16,40 @@ import (
 const (
 	AzurePriceSearchFilter = `serviceName eq 'Virtual Machines' and priceType eq 'Consumption'`
 	AzureMeterRegion       = `'primary'`
+
+	DefaultInstanceFamily = "General purpose"
```
Comment (Pokom) [inline pkg/azure/aks/price_store.go:20]: Ah this is interesting. TIL if you have a space between consts, `go-fmt` won't line up the equals signs. I think the intent here is to group consts?

File: pkg/azure/aks/price_store.go [hunk 4]
```diff
@@ -69,13 +96,26 @@ func NewPricingStore(parentContext context.Context, parentLogger *slog.Logger, s
 	return p, err
 }
 
-func (p *PriceStore) getPriceInfoFromVmInfo(vmInfo *VirtualMachineInfo) (float64, error) {
+func (p *PriceStore) getPriceBreakdownFromVmInfo(vmInfo *VirtualMachineInfo, price float64) *MachinePrices {
+	ratio, ok := cpuToCostRatio[vmInfo.MachineFamily]
+	if !ok {
+		p.logger.LogAttrs(p.context, slog.LevelInfo, "no ratio found for instance type, using default", slog.String("instanceType", vmInfo.MachineTypeSku), slog.String("instanceFamily", vmInfo.MachineFamily))
```
Comment (Pokom) [inline pkg/azure/aks/price_store.go:102]: nit: Split out each attribute to it's own line. It makes it much easier to read/grok.

File: pkg/azure/aks/price_store.go [hunk 5]
```diff
@@ -69,13 +96,26 @@ func NewPricingStore(parentContext context.Context, parentLogger *slog.Logger, s
...
+		ratio = cpuToCostRatio[DefaultInstanceFamily]
+	}
+
+	return &MachinePrices{
+		PricePerCore: price * ratio / float64(vmInfo.NumOfCores),
+		PricePerGiB:  (price * (1 - ratio) / float64(vmInfo.MemoryInMiB)) * 1024,
```
Comment (Pokom) [inline pkg/azure/aks/price_store.go:108]: Suggestion, make `1024` a const.

**Review by Pokom — APPROVED** [review]
👍🏻 Really good stuff! Ran locally and we had a conversation via slack about the startup time to hydrate the pricing store. Not a blocker for this PR, but something for us to consider when releasing is to have a liveness probe dependent on the pricing store being hydrated.

File: pkg/azure/aks/price_store.go (hunk 3 above)
Comment (logyball) [reply pkg/azure/aks/price_store.go:20]: updated to remove line between, and `go-fmt` lines up the equals sign!

File: pkg/azure/aks/price_store.go (hunk 4 above)
Comment (logyball) [reply pkg/azure/aks/price_store.go]: done

**Pull request merged**